File: coding-test-success/ch08/오픈채팅방.java
import java.util.*;

class Solution {
    public String[] solution(String[] record) {
        List<String> answer = new ArrayList<>();
        Map<String, String> m = new HashMap<>();
        
        // 1. record 처리: "Enter"와 "Change" 명령어 처리
        for (String line : record) {
            String[] cmd = line.split(" ");
            if(!cmd[0].equals("Leave"))
                m.put(cmd[1],cmd[2]);
        }
        
        // 2. 결과 메시지 생성
        for (String line : record) {
            String[] cmd = line.split(" ");
            if (cmd[0].equals("Enter")) {
                answer.add(m.get(cmd[1]) + "님이 들어왔습니다.");
            } else if (cmd[0].equals("Leave")) {
                answer.add(m.get(cmd[1]) + "님이 나갔습니다.");
            }
        }
        return answer.toArray(new String[0]);        
    }
}

File: coding-test-success/ch08/오픈채팅방.py
def solution(record):
    answer = []
    m = {}  # user_id : name

    # 1. record에서 정보를 처리
    for line in record:
        cmd = line.split(" ")
        if cmd[0] != "Leave":  # "Enter"와 "Change"일 경우에만 닉네임 업데이트
            m[cmd[1]] = cmd[2]

    # 2. 다시 record를 순차적으로 처리하여 출력 메시지 생성
    for info in record:
        cmd = info.split(" ")
        if cmd[0] == "Enter":
            answer.append(f"{m[cmd[1]]}님이 들어왔습니다.")
        elif cmd[0] == "Leave":
            answer.append(f"{m[cmd[1]]}님이 나갔습니다.")

    return answer
